### PR TITLE
New Adapter: Concert

### PR DESCRIPTION
**Pull request by BrettBlox**
Associated documentation PR: https://github.com/prebid/prebid.github.io/pull/5293

### CODE REVIEW

**Comment by gargcreation1992** [comment]
@BrettBlox PR checks are failing due to `expected.ImpIDs must contain at least one imp ID`. Please fix.

File: adapters/concert/concert.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	}
+	requestMap["ext"].(map[string]interface{})["adapterVersion"] = adapterVersion
+	requestMap["ext"].(map[string]interface{})["partnerId"] = bidderImpExt.PartnerId
+
+	requestJSON, err = json.Marshal(requestMap)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
```
Comment (Sonali-More-Xandr) [inline adapters/concert/concert.go:56]: Update `MakeRequests` to return `ImpIDs` as mentioned below :-
 
https://github.com/prebid/prebid-server/blob/763551963dd3208dde6f9bf35b31ef76cc83814f/adapters/pubmatic/pubmatic.go#L175-L182

Comment (BrettBlox) [reply adapters/concert/concert.go:60]: Done! Thank you for sharing the pubmatic example.

File: adapters/concert/concert.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
```
Comment (gargcreation1992) [inline adapters/concert/concert.go:87]: This is correct. But you could also use - 
```
if adapters.IsResponseStatusCodeNoContent(response) {
		return nil, nil
}

if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
		return nil, []error{err}
}
```

File: adapters/concert/concert.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	var errors []error
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			imp, _ := getImpByID(bid.ImpID, request.Imp)
+			bidType, err := getMediaTypeForBid(bid, imp)
```
Comment (gargcreation1992) [inline adapters/concert/concert.go:100]: Requesting you to please use Mtype. Prebid server expects the media type to be explicitly set in the adapter response. Therefore, recommends implementing a pattern where the adapter server sets the [MType](https://github.com/prebid/openrtb/blob/main/openrtb2/bid.go#L334) field in the response to accurately determine the media type for the impression.

File: adapters/concert/concert.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+// Builder builds a new instance of the Concert adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	bidderImpExt, err := getBidderExt(request.Imp[0])
```
Comment (gargcreation1992) [inline adapters/concert/concert.go:29]: Just re-checking - Are you sure to use just first impression and ignore others. If you wanna consider sending separate http request to your bidder server, you may choose to do so.

Comment (BrettBlox) [reply adapters/concert/concert.go:30]: Yes. As of now, this is only used for extracting the `partnerId` which should be the same across all Imp's.

File: adapters/concert/concert.go (hunk 3 above)
Comment (BrettBlox) [reply adapters/concert/concert.go]: Done! I have updated this to check MType first before falling back to checking the ext.Prebid field.

Comment (BrettBlox) [reply adapters/concert/concert.go]: Update: the fallback has been removed as it is no longer needed. We now will rely on the mtype property of the bid object.

File: static/bidder-info/concert.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+endpoint: "https://bids.concert.io/bids/openrtb"
+endpointCompression: gzip
+geoscope:
+  - global
+maintainer:
+  email: support@concert.io
```
Comment (Sonali-More-Xandr) [inline static/bidder-info/concert.yaml:6]: @BrettBlox We have sent a verification email to `support@concert.io`. Please reply to that mail with 'Received' to complete the verification process.

Comment (BrettBlox) [reply static/bidder-info/concert.yaml:6]: @Sonali-More-Xandr We just responded to your email.

File: adapters/concert/concert.go (hunk 2 above)
Comment (gargcreation1992) [reply adapters/concert/concert.go:88]: @BrettBlox This doesn't seem to have been implemented.

Comment (BrettBlox) [reply adapters/concert/concert.go:88]: @gargcreation1992 Correct. Is this change a requirement? Your previous comment made it seem like a suggested alternative and that both options were valid.

Comment (gargcreation1992) [reply adapters/concert/concert.go:88]: It's not a requirement but I would highly recommend making this change.

Comment (BrettBlox) [reply adapters/concert/concert.go]: @gargcreation1992 done.

File: static/bidder-info/concert.yaml (hunk 5 above)
Comment (BrettBlox) [reply static/bidder-info/concert.yaml:6]: @Sonali-More-Xandr was our response to your email received?

Comment (gargcreation1992) [reply static/bidder-info/concert.yaml:6]: email received. thanks!

**Review by Sonali-More-Xandr — APPROVED** [review]

**Pull request merged**